<a href="https://colab.research.google.com/github/ninadcpa/DSC540/blob/main/Week_7_8_Exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DSC540 Week 7 and 8 Exercise 
@author Ninad Patkhedkar

In [1]:
# This code is written for mounting code folder checked out from github on google drive
# All assignment code will be maintained inside folder "DSC540"
# Code from book is inside folder "Data-Wrangling-with-Python"

from google.colab import drive
import sys
import os

drive.mount('/content/drive')
sys.path.append('/content/drive/MyDrive/Colab Notebooks/DSC540')
os.chdir('/content/drive/MyDrive/Colab Notebooks/DSC540')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


The four chapters you read these two weeks focus extensively on cleaning and transforming data.

I am choosing below dataset for this exercise

*  [So Much Data Candy, Seriously (Ng, 2017)](https://www.scq.ubc.ca/wp-content/uploads/2017/10/candyhierarchy2017.csv) 

I have downloaded the data and will use local file.


In [2]:
import numpy as np
import pandas as pd

# File was throwing some encoding error while trying to read with default utf-8 encoding
# Using latin-1 encoding to fix issue
# Missing values or spaces would be replaced by np.nan
na_values = [""," ",None,"Missing","NA"]
df=pd.read_csv("candyhierarchy2017.csv",encoding='latin-1',na_values=na_values)
df.head()

Internal ID Q1: GOING OUT?  ... Q12: MEDIA [Yahoo] Click Coordinates (x, y)
0     90258773            NaN  ...                NaN                      NaN
1     90272821             No  ...                NaN                 (84, 25)
2     90272829            NaN  ...                NaN                      NaN
3     90272840             No  ...                NaN                 (75, 23)
4     90272841             No  ...                NaN                 (70, 10)

[5 rows x 120 columns]

Perform 2 tasks out of below list

Chapter 7
* Filter out missing data
* Fill in missing data
* Remove duplicates
* Transform data using either mapping or a function
* Replace values
* Discretization and Binning
* Manipulate Strings
* Noticed that missing Values are automatically populated as NaN



In [3]:
df.shape

(2460, 120)

In [4]:
df.isna().sum()

Internal ID                    0
Q1: GOING OUT?               110
Q2: GENDER                    41
Q3: AGE                       84
Q4: COUNTRY                   64
                            ... 
Q12: MEDIA [Daily Dish]     2375
Q12: MEDIA [Science]        1098
Q12: MEDIA [ESPN]           2361
Q12: MEDIA [Yahoo]          2393
Click Coordinates (x, y)     855
Length: 120, dtype: int64

Internal ID value is unique. I will use Interanl ID value as index.


In [5]:
df.set_index(['Internal ID'],inplace=True)
df.head()

Q1: GOING OUT?  ... Click Coordinates (x, y)
Internal ID                 ...                         
90258773               NaN  ...                      NaN
90272821                No  ...                 (84, 25)
90272829               NaN  ...                      NaN
90272840                No  ...                 (75, 23)
90272841                No  ...                 (70, 10)

[5 rows x 119 columns]

There are lot of missing values.
We will delete all rows for which all values are missing i.e. NaN

In [6]:
df.dropna(axis=0,how='all',inplace=True)
df.shape

(2439, 119)

It shows that around 21 rows dropped as it didn't had any value populated.

Check if there are any duplicate records.

In [7]:
df.duplicated().sum()

15

Drop duplicates.

In [8]:
df.drop_duplicates(inplace=True)
df.shape

(2424, 119)

There are few columns for which more than 50% records have missing values i.e. NaN.

I don't think those columns would be of any use, hence dropping those columns.

In [9]:
# Dropping columns where more than 50% records are NaN
df.dropna(thresh=df.shape[0]*0.5,how='all',axis=1,inplace=True)
df.shape

(2424, 112)

In [10]:
# Dropping rows where more than 50% records are NaN
df.dropna(thresh=df.shape[1]*0.5,how='all',axis=0,inplace=True)
df.shape

(1781, 112)

Notice that 7 columns are dropped.

Country names are missing at some places and also in different cases like US, us and some places usa

In [11]:
df['Q4: COUNTRY'].unique()

array(['USA ', 'us', 'usa', nan, 'canada', 'Canada', 'Us', 'US', 'Murica',
       'USA', 'uk', 'United States', 'united states', 'Usa',
       'United staes', 'United Kingdom', 'United States of America',
       'UAE', 'United States ', 'canada ', 'United states', 'u.s.a.',
       'USAUSAUSA', 'america', 'france', 'United States of America ',
       'U.S.A.', 'Mexico', 'Canada ', 'united states of america',
       'US of A', 'UK', 'Unites States', 'The United States',
       'North Carolina ', 'Netherlands', 'Europe', 'Earth', 'U S', 'u.s.',
       'Costa Rica', 'The United States of America', 'cascadia',
       'insanity lately', 'Greece', 'USA? Hard to tell anymore..',
       "'merica", 'Pittsburgh', 'United State', 'France', 'U.S.', 'A',
       'Can', 'Canae', 'New York', 'Trumpistan', 'Ireland',
       'United Sates', 'Korea', 'Australia', 'California', 'Japan', 'USa',
       'I pretend to be from Canada, but I am really from the United States.',
       'Usa ', 'Iceland', 'Canada`'

There are too many 129 countries with varying names. I will just try to correct country name for US.

First changing all contry names to uppercase.

In [12]:
df['Q4: COUNTRY'] = df['Q4: COUNTRY'].str.upper()

In [13]:
df['Q4: COUNTRY'].unique()

array(['USA ', 'US', 'USA', nan, 'CANADA', 'MURICA', 'UK',
       'UNITED STATES', 'UNITED STAES', 'UNITED KINGDOM',
       'UNITED STATES OF AMERICA', 'UAE', 'UNITED STATES ', 'CANADA ',
       'U.S.A.', 'USAUSAUSA', 'AMERICA', 'FRANCE',
       'UNITED STATES OF AMERICA ', 'MEXICO', 'US OF A', 'UNITES STATES',
       'THE UNITED STATES', 'NORTH CAROLINA ', 'NETHERLANDS', 'EUROPE',
       'EARTH', 'U S', 'U.S.', 'COSTA RICA',
       'THE UNITED STATES OF AMERICA', 'CASCADIA', 'INSANITY LATELY',
       'GREECE', 'USA? HARD TO TELL ANYMORE..', "'MERICA", 'PITTSBURGH',
       'UNITED STATE', 'A', 'CAN', 'CANAE', 'NEW YORK', 'TRUMPISTAN',
       'IRELAND', 'UNITED SATES', 'KOREA', 'AUSTRALIA', 'CALIFORNIA',
       'JAPAN',
       'I PRETEND TO BE FROM CANADA, BUT I AM REALLY FROM THE UNITED STATES.',
       'ICELAND', 'CANADA`', 'SCOTLAND', 'DENMARK', 'FRANCE ',
       'SWITZERLAND', 'AHEM....AMERCA', 'SOUTH KOREA', 'NEW JERSEY',
       'UNITED STATED', 'GERMANY', 'UNITED STATSS', 'ENDLAND

In [14]:
df['Q4: COUNTRY'].unique().size

79

Now will correct name for country of USA

In [15]:
usa_names_map = {
    'USA': 'USA',
    'USA ': 'USA',
    'US': 'USA',
    'UNITED STATES': 'USA',
    'U.S.': 'USA',
    'UNITED STATES OF AMERICA': 'USA',
    'US OF A': 'USA',
    'U.S.A.': 'USA',
    'USAUSAUSA': 'USA',
    'THE UNITED STATES':'USA',
    'UNIED STATES': 'USA',
    'USA! USA! USA!':'USA',
    'USAA': 'USA',
    'AMERICA': 'USA',
    'UNHINGED STATES': 'USA',
    'THE UNITED STATES OF AMERICA': 'USA', 
    'UNITE STATES': 'USA',
    'UNITED SATES': 'USA',
    'UNITES STATES': 'USA',
    'UNITED STSTES': 'USA',
    'UNITED STATES ': 'USA',
    'UNITED STATES OF AMERICA ': 'USA',
    'U S': 'USA',
    'UNITED STATED': 'USA',
    'USA USA USA!!!!': 'USA'
}

df['Q4: COUNTRY'] = df['Q4: COUNTRY'].map(usa_names_map).fillna(df['Q4: COUNTRY'])

In [16]:
df['Q4: COUNTRY'].unique()

array(['USA', nan, 'CANADA', 'MURICA', 'UK', 'UNITED STAES',
       'UNITED KINGDOM', 'UAE', 'CANADA ', 'FRANCE', 'MEXICO',
       'NORTH CAROLINA ', 'NETHERLANDS', 'EUROPE', 'EARTH', 'COSTA RICA',
       'CASCADIA', 'INSANITY LATELY', 'GREECE',
       'USA? HARD TO TELL ANYMORE..', "'MERICA", 'PITTSBURGH',
       'UNITED STATE', 'A', 'CAN', 'CANAE', 'NEW YORK', 'TRUMPISTAN',
       'IRELAND', 'KOREA', 'AUSTRALIA', 'CALIFORNIA', 'JAPAN',
       'I PRETEND TO BE FROM CANADA, BUT I AM REALLY FROM THE UNITED STATES.',
       'ICELAND', 'CANADA`', 'SCOTLAND', 'DENMARK', 'FRANCE ',
       'SWITZERLAND', 'AHEM....AMERCA', 'SOUTH KOREA', 'NEW JERSEY',
       'GERMANY', 'UNITED STATSS', 'ENDLAND', 'ATLANTIS', 'MURRIKA',
       'ALASKA', 'N. AMERICA', 'USSA', 'TAIWAN', 'CHINA', 'SPAIN',
       'U.S. ', 'NARNIA', 'U S A', 'UNITED STATEA',
       'SUBSCRIBE TO DM4UZ3 ON YOUTUBE', "I DON'T KNOW ANYMORE",
       'FEAR AND LOATHING'], dtype=object)

In [17]:
df['Q4: COUNTRY'].unique().size

61

Oops , it looks like too many corrections required for fixing country name. I won't fix it anymore as its taking lot of time. I am not sure if there is any better way to do this.

On side note, I think survey form should have dropdown rather than open text field for country. This can help with standardization and cleaner data collection.

## Chapter 8

I will perform below 2 tasks 
* Create hierarchical index
* Reshape

Candy names asked were part of Question 6 in survey. All Candy's surveyed are marked with Columns name starting with "Q6 |". Will extract column names and rename with new name removing "Q6 |".

In [18]:
candy_names = {x:x.split('Q6 | ')[1] for x in df.columns if x.startswith('Q6 | ')}

In [19]:
#candy_names = [ {x:x.split('Q6 | ')[1]} for x in df.columns if x.startswith('Q6 | ')] 
#candy_names
df.rename(columns=candy_names,inplace=True)
df.reset_index(inplace=True)
#df.set_index(list(candy_names.values()),inplace=True)
#df.head()

Dropping records where any of the Gender and Age field have missing values.

In [20]:
df.dropna(subset=['Q2: GENDER','Q3: AGE'],inplace=True)
df.shape

(1743, 113)

### Created Hierachial Data using columns Gender and Age

In [21]:
df.set_index(['Q2: GENDER','Q3: AGE'],inplace=True)
df.head()

Internal ID  ... Click Coordinates (x, y)
Q2: GENDER Q3: AGE               ...                         
Male       44          90272821  ...                 (84, 25)
           40          90272840  ...                 (75, 23)
           23          90272841  ...                 (70, 10)
           33          90272854  ...                  (55, 5)
           40          90272858  ...                 (76, 24)

[5 rows x 111 columns]

In [22]:
new_index = pd.Index(list(candy_names.values()),name='candy')
new_df = df.reindex(columns=new_index)
new_df.head()

candy              100 Grand Bar  ... York Peppermint Patties
Q2: GENDER Q3: AGE                ...                        
Male       44                MEH  ...                 DESPAIR
           40                MEH  ...                 DESPAIR
           23                JOY  ...                     JOY
           33                JOY  ...                 DESPAIR
           40                JOY  ...                 DESPAIR

[5 rows x 103 columns]

### Reshaping data using stack() method

In [23]:
ldata = new_df.stack().reset_index().rename(columns={0: 'candy_liking'})
ldata[:10]

Q2: GENDER  ... candy_liking
0       Male  ...          MEH
1       Male  ...      DESPAIR
2       Male  ...          JOY
3       Male  ...          MEH
4       Male  ...      DESPAIR
5       Male  ...      DESPAIR
6       Male  ...      DESPAIR
7       Male  ...      DESPAIR
8       Male  ...      DESPAIR
9       Male  ...      DESPAIR

[10 rows x 4 columns]

## Chapter 10

I will perform following 2 tasks

* Grouping with Dicts/Series
* Cross Tabs

I will use the dataframe created earlier i.e. df

In [24]:
df.reset_index(inplace=True)
df.head()

Q2: GENDER Q3: AGE  ...  Q12: MEDIA [Science] Click Coordinates (x, y)
0       Male      44  ...                   1.0                 (84, 25)
1       Male      40  ...                   1.0                 (75, 23)
2       Male      23  ...                   1.0                 (70, 10)
3       Male      33  ...                   1.0                  (55, 5)
4       Male      40  ...                   1.0                 (76, 24)

[5 rows x 113 columns]

In [25]:
df.shape

(1743, 113)

In [26]:
# dropping records where either country or State/Province/City field have missing data
df.dropna(subset=['Q4: COUNTRY','Q5: STATE, PROVINCE, COUNTY, ETC'],inplace=True)
df.shape

(1723, 113)

Now will group by Country and City. 

For that purpose lets create a mapping of columns

In [27]:
mapping = {
    'Q4: COUNTRY': 'COUNTRY',
    'Q5: STATE, PROVINCE, COUNTY, ETC': 'LOCATION'
}

by_column = df.groupby(mapping, axis=1)
by_column

In [28]:
by_column.describe()


count unique         top  freq
COUNTRY  Q4: COUNTRY                       1723     56         USA  1490
LOCATION Q5: STATE, PROVINCE, COUNTY, ETC  1723    423  California   107

Key Highlights
* Total Survey results - 1723
* Number of unique countries from where survey resulst are collected - 56
* Highest participation was for country USA and total respondents were 1490
* Number of unique locations i.e. State/City from survey results are collected - 423
* Highest participation in survey was from state of California i.e. 107

### CrossTab

Now I would like to find count of each candy marked as JOY , MEH and DISPAIR.
I will use CrossTab function.

I also want to find most liked Candy.

For this analysis , I will use dataframe created in Chapter 8 Exercise i.e. "ldata"

In [29]:
ldata[:10]

Q2: GENDER  ... candy_liking
0       Male  ...          MEH
1       Male  ...      DESPAIR
2       Male  ...          JOY
3       Male  ...          MEH
4       Male  ...      DESPAIR
5       Male  ...      DESPAIR
6       Male  ...      DESPAIR
7       Male  ...      DESPAIR
8       Male  ...      DESPAIR
9       Male  ...      DESPAIR

[10 rows x 4 columns]

Now Running crosstab function over candy and candy_liking column.

In [30]:
pd.crosstab([ldata['candy']],ldata.candy_liking,margins=True)

candy_liking                                        DESPAIR  ...     All
candy                                                        ...        
100 Grand Bar                                            84  ...    1668
Abstained from M&M'ing.                                 681  ...    1490
Anonymous brown globs that come in black and or...     1055  ...    1681
Any full-sized candy bar                                 16  ...    1734
Black Jacks                                             778  ...    1466
...                                                     ...  ...     ...
Whatchamacallit Bars                                    278  ...    1604
White Bread                                            1417  ...    1662
Whole Wheat anything                                   1257  ...    1674
York Peppermint Patties                                 226  ...    1711
All                                                   56898  ...  171514

[104 rows x 4 columns]

Lets find most liked candy.

In [31]:
crosstab_df = pd.crosstab([ldata['candy']],ldata.candy_liking)
print(crosstab_df[crosstab_df['JOY'] == crosstab_df['JOY'].max()])

candy_liking              DESPAIR   JOY  MEH
candy                                       
Any full-sized candy bar       16  1517  201


It looks like participants were more focused on size of candy rather than favour of the candy. "Any full size candy bar" was mosted liked candy in the survey.

## Chapter 11

I will perform following tasks
* Convert between string and date time
* Convert timestamps to periods and back


The dataset file copied from site [So Much Data Candy, Seriously (Ng, 2017)](https://www.scq.ubc.ca/wp-content/uploads/2017/10/candyhierarchy2017.csv)  does not contain any date field. So I will be using older file i.e. 2016 survey file which I downloaded from Bellevue University site [Week 7 & 8 Datasets](http://content.bellevue.edu/cst/dsc/540/id/data-files/12-week/weeks-7-&-8.zip). The filename is BOING-BOING-CANDY-HIERARCHY-2016-SURVEY-Responses.xlsx

In [32]:
na_values = [""," ",None,"Missing","NA"]
df_2016= pd.read_excel("BOING-BOING-CANDY-HIERARCHY-2016-SURVEY-Responses.xlsx",na_values=na_values)
df_2016.head()

Timestamp  ...  [York Peppermint Patties] Ignore
0 2016-10-24 05:09:23.033  ...                               NaN
1 2016-10-24 05:09:54.798  ...                               NaN
2 2016-10-24 05:13:06.734  ...                               NaN
3 2016-10-24 05:14:17.192  ...                               NaN
4 2016-10-24 05:14:24.625  ...                               NaN

[5 rows x 123 columns]

Checking Data type of Timestamp column.

In [33]:
dataTypeObj = df_2016.dtypes['Timestamp']
print(dataTypeObj)

datetime64[ns]


Dataset is loaded in panda dataframe with Timestamp column as type datetime64.

Lets me add day of the week field as extra column in dataframe. I will extract that value from timestamp field.

In [34]:
day_of_week = df_2016['Timestamp'].apply(lambda x: x.strftime('%A'))
# Adding day_of_week as 2nd column in dataframe
df_2016.insert(loc=1,column='Day_of_Week',value=day_of_week)
df_2016.sample(5)

Timestamp  ...  [York Peppermint Patties] Ignore
623 2016-10-24 15:15:47.759  ...                               NaN
309 2016-10-24 08:50:08.365  ...                               NaN
648 2016-10-24 16:28:17.087  ...                               NaN
277 2016-10-24 08:22:42.380  ...                               NaN
910 2016-10-25 19:04:26.731  ...                               NaN

[5 rows x 124 columns]

In [37]:
print(f" Data collected is between date range {df_2016['Timestamp'].min()} - {df_2016['Timestamp'].max()}")

 Data collected is between date range 2016-10-24 05:09:23.033000 - 2016-10-30 17:06:45.660000


SO data is collected during 7 days for 10/24 to 10/30 in year 2016.

### Convert timestamps to periods and back

I will convert timestamp into periods i.e. Morning, Afternoo, Evening and Night.
Then I will find how many survey results collected in those periods.

In [39]:
session=pd.cut(df_2016.Timestamp.dt.hour,
               [0,6,12,18,23],
               labels=['Night','Morning','Afternoon','Evening'],
               include_lowest=True)
session.value_counts()

Morning      586
Afternoon    362
Night        218
Evening       93
Name: Timestamp, dtype: int64